keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+
+Today, the certificates.k8s.io API group provides a flexible, pluggable
+mechanism for workloads running within the cluster to request certificates.
+However, certificates only have meaning in relation to a trust anchor set.  Most
+signers in a Kubernetes cluster don't issue publicly-trusted certificates, but
+Kubernetes lacks a standardized mechanism for signers to distribute their trust
+anchors to workloads.
+
+A trust anchor is an X.509 certificate that forms a root of trust &mdash; if you are
+presented with a leaf certificate, you decide whether or not to trust it based
+on whether or not you can form a chain of intermediate certificates from the
+leaf to a trust anchor.  This might be referred to as a “root certificate” in
+other contexts, but nothing actually forces a trust anchor to be a **root**;
+your trust anchor might be an intermediate certificate.  Additionally, note that
+trust anchors are contextual; depending on the identity that a client or server
+presents to you, you might wish to select a different trust anchor set.
+
+This KEP introduces the certificates.k8s.io TrustAnchorSets cluster-scoped
+object, a vehicle for in-cluster certificate signers to communicate their trust
+anchors to workloads.
+
+Additionally, this KEP introduces a new trustAnchorSets kubelet projected volume
+source that gives workloads easy filesystem-based access to trust anchor sets
+that they might need.
+
+## Motivation
+
+In the absence of a standard mechanism for trust distribution, a few different
+approaches have emerged:
+
+* The trust anchors necessary for connecting to kube-apiserver are distributed
+  using a ConfigMap that is automatically injected into every namespace.
+
+* Kubernetes webhooks and aggregated API servers are configured with a static
+  trust anchor set embedded in their config (which complicates rotation).
+
+* * At least one signer implementation uses a [cluster-scoped
+  CRD](https://cloud.google.com/traffic-director/docs/security-proxyless-setup#:~:text=Save%20the%20following%20TrustConfig%20YAML%20configuration%20to%20tell%20your%20cluster%20how%20to%20trust%20the%20issued%20certificates%3A)
+  to distribute trust anchors to workloads.
+
+Each of these approaches has downsides:
+
+* Creating a ConfigMap in every namespace requires granting over-broad
+  permissions to the responsible controller (create ConfigMaps in **all**
+  namespaces).
+
+* Static trust anchor sets embedded directly in configuration complicate the job
+  of rotating the backing CA's roots of trust.
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for certificates/v1 signer implementations to publish their
+  relevant trust anchors using a TrustAnchorSets object.
```

Comment:
Question: although this goal mentions specifically `certificates/v` signers I guess there would be no reason why this would not be used by other entities that act as signers and need to distribute CAs? If the overall goal is to unify how trust roots (including for publicly issued certs) are distributed in clusters, would it perhaps make sense to word this goal so as to include different types of signers?
(I'm specifically thinking about cert-manger where we woudl be interested to adopt this mechanism in our issuers (especially if, in the future, there is also a namespaced alternative of the new type))

Reply:
That makes sense, and I can update this language.  Does cert-manager not use other certificates/v1 stuff like CertificateSigningRequests?

Reply:
Updated.

Reply:
Thank you
> Does cert-manager not use other certificates/v1 stuff like CertificateSigningRequests

It can optionally use `CertificateSigningRequest`s, but by default uses our custom type `CertificateRequest` - the reason for not migrating is that `CertificateSigningRequest`s are not namespaced